process/blockProcessor.go

```diff
@@ -104,3 +105,53 @@ func (bp *blockProcessor) getObserversOrFullHistoryNodes(shardID uint32) ([]*dat
 
 	return bp.proc.GetObservers(shardID)
 }
+
+// GetHyperBlockByHash returns the hyperblock by hash
+func (bp *blockProcessor) GetHyperBlockByHash(hash string) (*data.HyperblockApiResponse, error) {
+	builder := &hyperblockBuilder{}
+
+	metaBlockResponse, err := bp.GetBlockByHash(core.MetachainShardId, hash, true)
+	if err != nil {
+		return nil, err
+	}
+
+	metaBlock := metaBlockResponse.Data.Block
+	builder.addMetaBlock(&metaBlock)
+
+	for _, notarizedBlock := range metaBlock.NotarizedBlocks {
+		shardBlockResponse, err := bp.GetBlockByHash(notarizedBlock.Shard, notarizedBlock.Hash, true)
+		if err != nil {
+			return nil, err
+		}
+
+		builder.addShardBlock(&shardBlockResponse.Data.Block)
+	}
+
+	hyperblock := builder.build()
+	return data.NewHyperblockApiResponse(hyperblock), nil
+}
+
+// GetHyperBlockByNonce returns the hyperblock by nonce
+func (bp *blockProcessor) GetHyperBlockByNonce(nonce uint64) (*data.HyperblockApiResponse, error) {
+	builder := &hyperblockBuilder{}
+
+	metaBlockResponse, err := bp.GetBlockByNonce(core.MetachainShardId, nonce, true)
+	if err != nil {
+		return nil, err
+	}
+
+	metaBlock := metaBlockResponse.Data.Block
+	builder.addMetaBlock(&metaBlock)
+
+	for _, notarizedBlock := range metaBlock.NotarizedBlocks {
+		shardBlockResponse, err := bp.GetBlockByNonce(notarizedBlock.Shard, notarizedBlock.Nonce, true)
```

Comment:
I think there's no need to get the shard block by nonce, as that would imply two storage accesses
1. Nonce to Header hash
2. header hash to header

But if we got the metablock by nonce, it means we already have access to the notarized blocks header hashes, so I would do getBlockByHash instead for the notarized blocks, as it should be faster.

Reply:
Fixed.